xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -365,3 +412,132 @@ func (b *cdsBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Sub
 func (b *cdsBalancer) Close() {
 	b.closed.Fire()
 }
+
+// ccWrapper wraps the balancer.ClientConn that was passed in to the CDS
+// balancer during creation and intercepts the NewSubConn() call from the child
+// policy. Other methods of the balancer.ClientConn interface are not overridden
+// and hence get the original implementation.
+type ccWrapper struct {
+	balancer.ClientConn
+	xdsCredsInUse bool
+
+	// This mutex protects all fields below.
+	mu        sync.Mutex
+	xdsHI     *xds.HandshakeInfo
+	xdsClient xdsClientInterface
+	// The certificate providers are cached here to that they can be closed when
+	// a new provider is to be created.
+	cachedRoot     certprovider.Provider
+	cachedIdentity certprovider.Provider
+}
+
+// NewSubConn handles intercepts attempts create a new SubConn from the child
+// policy and adds an address attribute which provides all information required
+// by the xdsCreds handshaker to perform the TLS handshake.
+func (ccw *ccWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	if !ccw.xdsCredsInUse || ccw.xdsHI == nil {
+		return ccw.ClientConn.NewSubConn(addrs, opts)
+	}
+
+	newAddrs := make([]resolver.Address, len(addrs))
+	for i, addr := range addrs {
+		newAddrs[i] = xds.SetHandshakeInfo(addr, ccw.xdsHI)
+	}
+	return ccw.ClientConn.NewSubConn(newAddrs, opts)
+}
+
+// setClient updates the xdsClient within the wrapper.
+func (ccw *ccWrapper) setClient(c xdsClientInterface) {
+	ccw.mu.Lock()
+	ccw.xdsClient = c
+	ccw.mu.Unlock()
+}
+
+// handleSecurityConfig processes the security configuration received from the
+// management server, creates appropriate certificate provider plugins, and
+// updates the HandhakeInfo which is added as an address attribute in
+// NewSubConn() calls.
+func (ccw *ccWrapper) handleSecurityConfig(config *xdsclient.SecurityConfig) error {
```

Comment:
Could this be done in the parent `cdsBalancer` instead of here?  Then we wouldn't need an xdsClient in this `ccw` and the only thing the `ccw` would do is attach `xdsHI` in the address attributes.

Also, we just wouldn't use a wrapper at all if `xdsCredsInUse` was not true; so we don't need an extra check for `!= nil` - it'd be just:

```go
type ccWrapper struct {
	balancer.ClientConn
	xdsHI *xds.HandshakeInfo
}

func (ccw *ccWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
	newAddrs := make([]resolver.Address, len(addrs))
	for i, addr := range addrs {
		newAddrs[i] = xds.SetHandshakeInfo(addr, ccw.xdsHI)
	}
	return ccw.ClientConn.NewSubConn(newAddrs, opts)
}
```

Reply:
Done. Thanks for the suggestion. Looks much better now.